### PR TITLE
Feature 45 As a developer I want to have a service which will save and share video files

### PR DESCRIPTION
**Pull request by NickRost**

### CODE REVIEW

File: backend/Recodo.API.BLL/Recodo.API.BLL.csproj [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (Jyuart) [inline backend/Recodo.API.BLL/Recodo.API.BLL.csproj]: Please, rename this project to Recodo.FileAPI.BLL or something like that to mark that it's a part of a separate API

File: backend/Recodo.API.BLL/Services/AzureBlobService.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    public class AzureBlobService : IAzureBlobService
+    {
+        private readonly IAzureBlobConnectionFactory _azureBlobConnectionFactory;
+
+        public AzureBlobService(IAzureBlobConnectionFactory azureBlobConnectionFactory)
+        {
+            _azureBlobConnectionFactory = azureBlobConnectionFactory;
+        }
+
+		public async Task<IEnumerable<Uri>> ListAsync()
```
Comment (Jyuart) [inline backend/Recodo.API.BLL/Services/AzureBlobService.cs]: Fix indentation

File: backend/Recodo.API.BLL/Services/AzureBlobService.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+		public async Task UploadAsync(int id, IFormFile files)
+		{
+			var blobContainer = await _azureBlobConnectionFactory.GetBlobContainer();
+			var blob = blobContainer.GetBlockBlobReference(id.ToString());
+
+			using (var fileStream = files.OpenReadStream())
+			{
+				await blob.UploadFromStreamAsync(fileStream);
+			}
```
Comment (Jyuart) [inline backend/Recodo.API.BLL/Services/AzureBlobService.cs]:
```suggestion
            await using var fileStream = files.OpenReadStream();
            await blob.UploadFromStreamAsync(fileStream);
```
?

File: backend/Recodo.BLL/Services/BlobService.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Recodo.BLL.Services
+{
+    internal class BlobService
+    {
+        private readonly BlobServiceClient _blobServiceClient;
+        const string BlobContainerName = "recodo-filecontainer";
```
Comment (Jyuart) [inline backend/Recodo.BLL/Services/BlobService.cs]: Move it to appsettings.json

File: backend/Recodo.BLL/Services/BlobService.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Recodo.BLL.Services
+{
+    internal class BlobService
+    {
+        private readonly BlobServiceClient _blobServiceClient;
+        const string BlobContainerName = "recodo-filecontainer";       
+        private readonly CloudBlobContainer blobContainer;
```
Comment (Jyuart) [inline backend/Recodo.BLL/Services/BlobService.cs]:
```suggestion
        private readonly CloudBlobContainer _blobContainer;
```
Keep fields naming consistency

File: backend/Recodo.BLL/Services/BlobService.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+using Microsoft.WindowsAzure.Storage.Blob;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Recodo.BLL.Services
+{
+    internal class BlobService
```
Comment (Jyuart) [inline backend/Recodo.BLL/Services/BlobService.cs]: What's the purpose of this service anyway?

File: backend/Recodo.BlobAPI/Controllers/BlobExplorerController.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        [HttpGet("list")]
+        public async Task<IActionResult> GetBlobs()
+        {
+            return Ok(await _blobService.ListAsync());
+        }
+
+        [HttpGet]
+        public async Task<FileStreamResult> GetFile(int id)
+        {      
+            return File(await _blobService.DownloadAsync(id), "application/mp4", id.ToString()+".mp4");
```
Comment (Jyuart) [inline backend/Recodo.BlobAPI/Controllers/BlobExplorerController.cs]: This should happen in the BLL layer. Just pass the data in the controller
And please use string extrapolation

**Review by Jyuart — CHANGES_REQUESTED** [review]

File: backend/Recodo.API.BLL/Services/FIleService.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+		public FileService(IAzureBlobConnectionFactory azureBlobConnectionFactory, IRequestService requestService)
+        {
+            _azureBlobConnectionFactory = azureBlobConnectionFactory;
+			_requestService = requestService;
+        }
+
+		public async Task<IEnumerable<Uri>> ListAsync()
+		{
+			var blobContainer = await _azureBlobConnectionFactory.GetBlobContainer();
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API.BLL/Services/FIleService.cs]: Tabulation

File: backend/Recodo.API.BLL/Services/FIleService.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        private readonly IAzureBlobConnectionFactory _azureBlobConnectionFactory;
+		private readonly IRequestService _requestService;
+
+		public FileService(IAzureBlobConnectionFactory azureBlobConnectionFactory, IRequestService requestService)
+        {
+            _azureBlobConnectionFactory = azureBlobConnectionFactory;
+			_requestService = requestService;
+        }
+
+		public async Task<IEnumerable<Uri>> ListAsync()
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API.BLL/Services/FIleService.cs]: Do you need this?

File: backend/Recodo.API.BLL/Services/RequestService.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Recodo.FIle.BLL.Services
+{
+    public class RequestService : IRequestService
+    {
+        private static HttpClient Client { get; set; } = new HttpClient();
+
+        private static string BaseUrl = new string(@"https://localhost:44316/api/");
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API.BLL/Services/RequestService.cs]: To appsettings

File: backend/Recodo.API.BLL/Services/RequestService.cs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+        public async Task<string> SendSaveRequest(string token)
+        {
+            Client.DefaultRequestHeaders.Add(HeaderNames.Authorization, token);
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.API.BLL/Services/RequestService.cs:28]: static HttpCLient + DefaultRequestHeaders? Can you make HttpClient scoped at least?

File: backend/Recodo.BLL/Services/FileService.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        private readonly RecodoDbContext _context;
+
+        public FileService(RecodoDbContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<int> SaveVideo(int authorId)
+        {
+            Video newVideo = new Video();
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/FileService.cs]: This way please 
            var newVideo = new Video() {
            IsSaving = true,
            AuthorId = authorId,
            ...
            }

File: backend/Recodo.BLL/Services/FileService.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            newVideo.IsSaving = true;
+            newVideo.AuthorId = authorId;
+            newVideo.CreatedAt = DateTime.Now;
+            newVideo.Name = "Video";
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/FileService.cs]: Set default name in format Video_Date_Time. As an example Video_23022022_2008

File: backend/Recodo.BLL/Services/FileService.cs [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+            return newVideo.Id;
+        }
+
+        public async Task FinishLoadingFile(int id)
+        {
+            var video = await _context.Videos.FirstOrDefaultAsync(x => x.Id == id);
+            if (video == null)
+            {
+                throw new Exception("Invalid id");
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/FileService.cs]: Video with such identifier is not found. The same at the bottom

File: backend/Recodo.BlobAPI/Controllers/FileController.cs [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        public FileController(IFileService blobService)
+        {
+            _blobService = blobService;
+        }
+
+        [HttpGet("list")]
+        public async Task<IActionResult> GetBlobs()
+        {
+            return Ok(await _blobService.ListAsync());
+        }
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BlobAPI/Controllers/FileController.cs]: We don't need it, please create secure logic

File: backend/Recodo.BlobAPI/Controllers/FileController.cs [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+        [HttpPost]
+        [RequestFormLimits(MultipartBodyLengthLimit = 157286400)]
+        public async Task<IActionResult> UploadFile(IFormFile file)
+        {
+            var accessToken = Request.Headers[HeaderNames.Authorization];
+            if (await _blobService.UploadAsync(file, accessToken.FirstOrDefault()))
+            {
+                return Ok();
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BlobAPI/Controllers/FileController.cs]: Can we return id when uploading is finished?

File: backend/Recodo.BlobAPI/web.config [hunk 17]
```diff
@@ -0,0 +1,10 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+	<system.webServer>
+		<security>
+			<requestFiltering>
+				<requestLimits maxAllowedContentLength="157286400"/>
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BlobAPI/web.config]: Increase to 500MB

File: backend/Recodo.API.BLL/Services/RequestService.cs (hunk 11 above)
Comment (NickRost) [reply backend/Recodo.API.BLL/Services/RequestService.cs:28]: Scoped this service in startup? Or I should declare HttpClient inside a method?

Comment (Anatolii-Ivanov) [reply backend/Recodo.API.BLL/Services/RequestService.cs:28]: Scoped in startup

File: backend/Recodo.BLL/Services/FileService.cs [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            var video = await _context.Videos.FirstOrDefaultAsync(x => x.Id == id);
+            if (video == null)
+            {
+                throw new Exception("Video with such identifier is not found.");
+            }
+
+            video.IsSaving = false;
+        }
+        public async Task GetFile(int userId, int videoId)
```
Comment (Anatolii-Ivanov) [inline backend/Recodo.BLL/Services/FileService.cs]: What about returning something?

**Review by Anatolii-Ivanov — APPROVED** [review]

**Review by Jyuart — APPROVED** [review]

**Pull request merged**